File: src/client/java/io/github/ricciow/format/SpecialFunctions.java
package io.github.ricciow.format;

import io.github.ricciow.PridgeClient;
import io.github.ricciow.config.PridgeConfig;
import io.github.ricciow.util.ColorCode;
import io.github.ricciow.util.STuF;
import io.github.ricciow.util.TextParser;
import net.minecraft.text.*;
import net.minecraft.util.Formatting;

import java.net.URI;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SpecialFunctions {

    @FunctionalInterface
    public interface SpecialFunction {
        FormatResult run(String originalText, Matcher matcher);
    }

    private static final Map<String, SpecialFunction> registry = new HashMap<>();

    private static PridgeConfig CONFIG;
    private static final Pattern LINK_PATTERN = Pattern.compile("\\[LINK\\]\\((l\\$[^)]+)\\)");

    static {
        registry.put("discord", SpecialFunctions::discordHandler);
        registry.put("contest1", SpecialFunctions::contest1Handler);
        registry.put("contest2", SpecialFunctions::contest2Handler);
        registry.put("contest3", SpecialFunctions::contest3Handler);
        registry.put("contest4", SpecialFunctions::contest4Handler);
        registry.put("bestiary", SpecialFunctions::bestiaryHandler);
        registry.put("bestiary2", SpecialFunctions::bestiary2Handler);
        registry.put("collection", SpecialFunctions::collectionHandler);
    }

    /**
     * Gets a function from the corresponding name
     * @param functionName Name of the function
     * @return An optional containing a single function to be run with .run()
     */
    public static Optional<SpecialFunction> get(String functionName) {
        return Optional.ofNullable(registry.get(functionName));
    }

    /**
     * Runs a function with the corresponding name
     * @param functionName Name of the function to be run
     * @param matcher RegExp matcher to be passed through the function
     * @return String containing the result of the function or null if there isn't a function
     */
    public static FormatResult run(String functionName, String originalText, Matcher matcher) {
        SpecialFunction function = registry.get(functionName);

        if(function != null) {
            return function.run(originalText, matcher);
        }

        return null;
    }

    //<editor-fold desc="Helper Methods">

    /**
     * Formats a time value, returning an empty string if the value is 0.
     * @param timeValue The numeric value of the time unit.
     * @param suffix The suffix to append (e.g., "h", "m", "s").
     * @return A formatted string like " 10h" or an empty string.
     */
    private static String timeFunc(int timeValue, String suffix) {
        if (timeValue == 0) {
            return "";
        }
        return " " + timeValue + suffix;
    }

    /**
     * Capitalizes the first letter of a string.
     * @param str The string to capitalize.
     * @return The capitalized string.
     */
    private static String capitalizeFirstLetter(String str) {
        if (str == null || str.isEmpty()) {
            return str;
        }
        return str.substring(0, 1).toUpperCase() + str.substring(1);
    }

    /**
     * Placeholder for a function that formats hyperlinks.
     * In a real mod, this would involve creating chat components with click events.
     * @return A formatted string or null if no links are found.
     */
    private static Text formatLink(String originalText, Matcher matcher) {
        if(CONFIG.linkCategory.enabled) {
            //Decode URL
            String URL = STuF.decode(matcher.group(1));

            //Make Click thing
            ClickEvent openUrlEvent = new ClickEvent.OpenUrl(URI.create(URL));
            Style linkStyle = Style.EMPTY.withClickEvent(openUrlEvent);
            MutableText result = TextParser.parse(CONFIG.linkCategory.representation);
            result.setStyle(linkStyle);

            return result;
        }

        return TextParser.parse(originalText);

    }

    private static PridgeConfig getConfig() {
        if(CONFIG == null) {
            CONFIG = PridgeClient.getConfig();
        }
        return CONFIG;
    }

    private static FormatResult contest1Handler(String originalText, Matcher matcher) {
        String crop = matcher.group(1);
        int hours = Integer.parseInt(matcher.group(2));
        int minutes = Integer.parseInt(matcher.group(3));
        int seconds = Integer.parseInt(matcher.group(4));

        String hoursStr = timeFunc(hours, "h");
        String minutesStr = timeFunc(minutes, "m");
        String secondsStr = timeFunc(seconds, "s");

        String result = String.format("&eNext %s contest in%s%s%s", crop, hoursStr, minutesStr, secondsStr);

        return new FormatResult(originalText, result, false, true);
    }

    private static FormatResult contest2Handler(String originalText, Matcher matcher) {
        String crop1 = matcher.group(1);
        String crop2 = matcher.group(2);
        String crop3 = matcher.group(3);
        // Groups 4 (minutesActive) and 5 (secondsActive) are unused in the final format string.
        String nextCrop = matcher.group(6);
        int hours = Integer.parseInt(matcher.group(7));
        int minutes = Integer.parseInt(matcher.group(8));
        int seconds = Integer.parseInt(matcher.group(9));

        String hoursStr = timeFunc(hours, "h");
        String minutesStr = timeFunc(minutes, "m");
        String secondsStr = timeFunc(seconds, "s");

        String result = String.format("\n &a&lActive Contest\n &6%s, %s, %s\n&eNext %s contest in&f%s%s%s",
                        crop1, crop2, crop3, nextCrop, hoursStr, minutesStr, secondsStr);

        return new FormatResult(originalText, result, false, true);
    }

    private static FormatResult contest3Handler(String originalText, Matcher matcher) {
        String crop1 = matcher.group(1);
        String crop2 = matcher.group(2);
        String crop3 = matcher.group(3);
        // Groups 4 (minutesActive) and 5 (secondsActive) are unused.
        String crop4 = matcher.group(6);
        String crop5 = matcher.group(7);
        String crop6 = matcher.group(8);
        int hours = Integer.parseInt(matcher.group(9));
        int minutes = Integer.parseInt(matcher.group(10));
        int seconds = Integer.parseInt(matcher.group(11));

        String hoursStr = timeFunc(hours, "h");
        String minutesStr = timeFunc(minutes, "m");
        String secondsStr = timeFunc(seconds, "s");

        String result = String.format("\n &a&lActive Contest\n &6%s, %s, %s\n &e&lNext: \n &6%s, %s, %s\n &eIn&f%s%s%s",
                        crop1, crop2, crop3, crop4, crop5, crop6, hoursStr, minutesStr, secondsStr);

        return new FormatResult(originalText, result, false, true);
    }

    private static FormatResult contest4Handler(String originalText, Matcher matcher) {
        String crop1 = matcher.group(1);
        String crop2 = matcher.group(2);
        String crop3 = matcher.group(3);
        int hours = Integer.parseInt(matcher.group(4));
        int minutes = Integer.parseInt(matcher.group(5));
        int seconds = Integer.parseInt(matcher.group(6));

        String hoursStr = timeFunc(hours, "h");
        String minutesStr = timeFunc(minutes, "m");
        String secondsStr = timeFunc(seconds, "s");

        String result = String.format("&e&lNext:\n &6%s, %s, %s\n &eIn&f%s%s%s",
                        crop1, crop2, crop3, hoursStr, minutesStr, secondsStr);

        return new FormatResult(originalText, result, false, true);
    }

    private static FormatResult bestiaryHandler(String originalText, Matcher matcher) {
        int maxPerPage = getConfig().botCategory.getLineCount();

        String bestiary = matcher.group(1);
        String user = matcher.group(2);
        String profile = matcher.group(3);
        String message = matcher.group(4);

        // A single, more robust pattern to capture all data points at once
        Pattern bestiaryDataPattern = Pattern.compile("(\\w[\\w\\s]*?) (\\d+)/(\\d+)(?: \\(([\\d.]+)\\))?");

        Text prefix = TextParser.parse(String.format("\n &6&l%s bestiary - &f&l%s (&f&l%s)&6&l\n ", bestiary, user, profile));
        List<Text> pages = new ArrayList<>();
        StringBuilder currentPageContent = new StringBuilder();
        int entriesOnPage = 0;

        // Use the single pattern on the whole message
        Matcher entryMatcher = bestiaryDataPattern.matcher(message);
        while (entryMatcher.find()) {
            if (entriesOnPage == maxPerPage) {
                pages.add(TextParser.parse(currentPageContent.toString()));
                currentPageContent.setLength(0); // More efficient than new StringBuilder()
                entriesOnPage = 0;
            }

            String name = entryMatcher.group(1).trim();
            String current = entryMatcher.group(2);
            String max = entryMatcher.group(3);
            String valueStr = entryMatcher.group(4); // Can be null

            if(entriesOnPage != 0) {
                currentPageContent.append("\n");
            }

            currentPageContent.append(String.format(" &e&l%s &f&l%s&e&l/&f&l%s", name, current, max));

            if (valueStr != null) {
                double value = Double.parseDouble(valueStr);
                String color;
                if (value > 1)         color = ColorCode.GREEN.getCode();
                else if (value > 0.75) color = ColorCode.YELLOW.getCode();
                else if (value > 0.5)  color = ColorCode.GOLD.getCode();
                else if (value > 0.25) color = ColorCode.RED.getCode();
                else                   color = ColorCode.DARK_RED.getCode();
                currentPageContent.append(String.format(" &e&l(%s&l%s&e&l)", color, valueStr));
            } else {
                currentPageContent.append(" &e&l(&a&lPro&e&l)");
            }
            entriesOnPage++;
        }

        // Add the final page if it has any content
        if (currentPageContent.length() > 0) {
            if(!pages.isEmpty()) {
                while (entriesOnPage < maxPerPage) {
                    currentPageContent.append("\n");
                    entriesOnPage += 1;
                }
            }
            pages.add(TextParser.parse(currentPageContent.toString()));
        }

        List<Text> titles = new ArrayList<>();
        for(int i = 1; i <= pages.size(); i++) {
            titles.add(TextParser.parse(String.format("&6Page (%s/%s)", i, pages.size())));
        }

        return new FormatResult(pages, titles, TextColor.fromFormatting(Formatting.DARK_AQUA), TextColor.fromFormatting(Formatting.GRAY), prefix, false, true);
    }

    private static FormatResult bestiary2Handler(String originalText, Matcher matcher) {
        String mob = matcher.group(1);
        String user = matcher.group(2);
        String profile = matcher.group(3);
        String num = matcher.group(4);

        String str = (Integer.parseInt(num) > 0) ? "&a&lPro" : "&4&l0";

        String result = String.format("&f&l%s (&f&l%s)&6&l:\n &6&l%s - &f&l%s&e&l/&f&l0 &e&l(%s&e&l)",
                        user, profile, mob, num, str);

        return new FormatResult(originalText, result, false, true);
    }

    private static FormatResult collectionHandler(String originalText, Matcher matcher) {
        int maxPerPage = getConfig().botCategory.getLineCount();

        String skill = matcher.group(1);
        String user = matcher.group(2);
        String profile = matcher.group(3);
        String message = matcher.group(4);

        // 1. Replaced the two patterns with a single, more direct one.
        // This pattern captures all necessary groups in one pass.
        Pattern dataPattern = Pattern.compile("([\\w\\s]*) (\\d+)/(\\d+) \\(([^)]+)\\)");

        Text prefix = TextParser.parse(String.format("\n &6&l%s collections - &f&l%s (&f&l%s)&6&l\n ",
                capitalizeFirstLetter(skill), user, profile));

        List<Text> pages = new ArrayList<>();
        StringBuilder currentPageContent = new StringBuilder();
        int entriesOnPage = 0;

        // 2. The loop is now simpler, with no nested matching.
        Matcher entryMatcher = dataPattern.matcher(message);
        while (entryMatcher.find()) {
            if (entriesOnPage == maxPerPage) {
                pages.add(TextParser.parse(currentPageContent.toString()));
                currentPageContent.setLength(0);
                entriesOnPage = 0;
            }

            // 3. All data is retrieved directly from the single 'entryMatcher'.
            String name = entryMatcher.group(1).trim();
            String current = entryMatcher.group(2);
            String max = entryMatcher.group(3);
            String progress = entryMatcher.group(4); // e.g., "1,234" or "1,234/5,678"

            if (entriesOnPage != 0) {
                currentPageContent.append("\n");
            }

            int currentVal = Integer.parseInt(current);
            int maxVal = Integer.parseInt(max);
            int medianVal = Math.floorDiv(maxVal, 4);

            ColorCode numberColor;
            if(currentVal == maxVal) {
                numberColor = ColorCode.GREEN;
            }
            else if (currentVal > medianVal) {
                numberColor = ColorCode.GOLD;
            }
            else {
                numberColor = ColorCode.RED;
            }

            String formattedProgress = progress.replace("/", "&e&l/&f&l");

            currentPageContent.append(String.format(" &e&l%s %s&l%s&e&l/%s&l%s &e&l(&f&l%s&e&l)",
                    name, numberColor.getCode(), current, numberColor.getCode(), max, formattedProgress));

            entriesOnPage++;
        }

        // This logic for handling the final page remains the same.
        if (currentPageContent.length() > 0) {
            if(!pages.isEmpty()) {
                while (entriesOnPage < maxPerPage) {
                    currentPageContent.append("\n");
                    entriesOnPage++;
                }
            }
            pages.add(TextParser.parse(currentPageContent.toString()));
        }

        // This logic for generating titles remains the same.
        List<Text> titles = new ArrayList<>();
        for(int i = 1; i <= pages.size(); i++) {
            titles.add(TextParser.parse(String.format("&6Page (%s/%s)", i, pages.size())));
        }

        // The return statement is the same.
        return new FormatResult(pages, titles, TextColor.fromFormatting(Formatting.DARK_AQUA), TextColor.fromFormatting(Formatting.GRAY), prefix, false, true);
    }

    private static FormatResult discordHandler(String originalText, Matcher matcher) {
        String user = matcher.group(1);
        String message = matcher.group(2);
        String userName;

        // This logic handles cases where the message contains ": " and the regex captures it as part of the user.
        // It correctly reassembles the user and message parts.
        if (user.contains(": ")) {
            String[] parts = user.split(": ");
            userName = parts[0];
            StringBuilder messageBuilder = new StringBuilder();
            for (int i = 1; i < parts.length; i++) {
                messageBuilder.append(parts[i]).append(": ");
            }
            messageBuilder.append(message);
            message = messageBuilder.toString();
        } else {
            userName = user;
        }

//        checkForSounds(message);

        MutableText finalMessage = TextParser.parse(
                getConfig().discordCategory.name_color.get().getCode() +
                userName +
                getConfig().discordCategory.message_color.get().getCode() + ": ");

        String[] parts = message.split(" ");
        boolean starting = true;
        for (String part : parts) {

            if (part.isEmpty()) {
                continue;
            }

            Matcher partMatcher = LINK_PATTERN.matcher(part);
            Text formattedPart;
            //Treat links
            if(partMatcher.matches()) {
                formattedPart = formatLink(part, partMatcher);
            }
            else {
                formattedPart = TextParser.parse(getConfig().discordCategory.message_color.get().getCode() + part);
            }

            //Add to the end of the message;
            if(!starting) {
                finalMessage.append(" ");
            }
            else {
                starting = false;
            }
            finalMessage.append(formattedPart);
        }

        return new FormatResult(originalText, finalMessage, true, false);
    }
}

File: src/main/kotlin/io/github/ricciow/format/SpecialFunctions.kt
package io.github.ricciow.format

import io.github.ricciow.Pridge.CONFIG_I
import io.github.ricciow.util.ColorCode
import io.github.ricciow.util.PridgeLogger
import io.github.ricciow.util.TextParser.parse
import io.github.ricciow.util.UrlFormatter
import io.github.ricciow.util.toText
import net.minecraft.text.ClickEvent.OpenUrl
import net.minecraft.text.HoverEvent.ShowText
import net.minecraft.text.Style
import net.minecraft.text.Text
import net.minecraft.text.TextColor
import net.minecraft.util.Formatting
import java.net.URI

object SpecialFunctions {
    private val registry = mutableMapOf<String, SpecialFunction>()

    private val LINK_PATTERN = Regex("\\[LINK]\\((l\\$[^)]+)\\)")

    fun initialize() {
        registry.put("discord", SpecialFunctions::discordHandler)
        registry.put("contest1", SpecialFunctions::contest1Handler)
        registry.put("contest2", SpecialFunctions::contest2Handler)
        registry.put("contest3", SpecialFunctions::contest3Handler)
        registry.put("contest4", SpecialFunctions::contest4Handler)
        registry.put("bestiary", SpecialFunctions::bestiaryHandler)
        registry.put("bestiary2", SpecialFunctions::bestiary2Handler)
        registry.put("collection", SpecialFunctions::collectionHandler)
    }

    /**
     * Gets a function from the corresponding name
     * @param functionName Name of the function
     * @return An optional containing a single function to be run with .run()
     */
    fun get(functionName: String): SpecialFunction? {
        return registry[functionName]
    }

    /**
     * Runs a function with the corresponding name
     * @param functionName Name of the function to be run
     * @param matcher RegExp matcher to be passed through the function
     * @return String containing the result of the function or null if there isn't a function
     */
    fun run(functionName: String, originalText: String, matcher: MatchResult): FormatResult? {
        val function = get(functionName) ?: return null
        return function.run(originalText, matcher)
    }

    //<editor-fold desc="Helper Methods">
    /**
     * Formats a time value, returning an empty string if the value is 0.
     * @param timeValue The numeric value of the time unit.
     * @param suffix The suffix to append (e.g., "h", "m", "s").
     * @return A formatted string like " 10h" or an empty string.
     */
    private fun timeFunc(timeValue: Int, suffix: String): String {
        if (timeValue == 0) return ""
        return " $timeValue$suffix"
    }

    /**
     * Capitalizes the first letter of a string.
     * @param str The string to capitalize.
     * @return The capitalized string.
     */
    private fun capitalizeFirstLetter(str: String?): String? {
        if (str == null || str.isEmpty()) return str
        return str.substring(0, 1).uppercase() + str.substring(1)
    }

    /**
     * Placeholder for a function that formats hyperlinks.
     * In a real mod, this would involve creating chat components with click events.
     * @return A formatted string or null if no links are found.
     */
    private fun formatLink(originalText: String, matcher: MatchResult): Text {
        if (CONFIG_I.linkCategory.enabled) {
            var representation = CONFIG_I.linkCategory.representation
            val group = matcher.groupValues[1]
            val url = try {
                UrlFormatter.decode(group)
            } catch (e: Exception) {
                PridgeLogger.error("Failed to decode URL: $group", e)
                representation = "&a&l[Failed to decode URL]"
                "https://github.com/Ricciow/Pridge-1.21.5/issues"
            }
            return parse(representation).apply {
                style = Style.EMPTY
                    .withClickEvent(OpenUrl(URI(url)))
                    .withHoverEvent(ShowText(url.toText()))
            }
        }

        return parse(originalText)
    }

    private fun contest1Handler(originalText: String, matcher: MatchResult): FormatResult {
        val crop = matcher.groupValues[1]
        val hours = matcher.groupValues[2].toInt()
        val minutes = matcher.groupValues[3].toInt()
        val seconds = matcher.groupValues[4].toInt()

        val hoursStr = timeFunc(hours, "h")
        val minutesStr = timeFunc(minutes, "m")
        val secondsStr = timeFunc(seconds, "s")

        return FormatResult("&eNext $crop contest in&f$hoursStr$minutesStr$secondsStr", botText = true)
    }

    private fun contest2Handler(originalText: String, matcher: MatchResult): FormatResult {
        val crop1 = matcher.groupValues[1]
        val crop2 = matcher.groupValues[2]
        val crop3 = matcher.groupValues[3]
        // Groups 4 (minutesActive) and 5 (secondsActive) are unused in the final format string.
        val nextCrop = matcher.groupValues[6]
        val hours = matcher.groupValues[7].toInt()
        val minutes = matcher.groupValues[8].toInt()
        val seconds = matcher.groupValues[9].toInt()

        val hoursStr = timeFunc(hours, "h")
        val minutesStr = timeFunc(minutes, "m")
        val secondsStr = timeFunc(seconds, "s")

        return FormatResult(
            "\n &a&lActive Contest\n &6$crop1, $crop2, $crop3\n&eNext $nextCrop contest in&f$hoursStr$minutesStr$secondsStr",
            botText = true
        )
    }

    private fun contest3Handler(originalText: String, matcher: MatchResult): FormatResult {
        val crop1 = matcher.groupValues[1]
        val crop2 = matcher.groupValues[2]
        val crop3 = matcher.groupValues[3]
        // Groups 4 (minutesActive) and 5 (secondsActive) are unused.
        val crop4 = matcher.groupValues[6]
        val crop5 = matcher.groupValues[7]
        val crop6 = matcher.groupValues[8]
        val hours = matcher.groupValues[9].toInt()
        val minutes = matcher.groupValues[10].toInt()
        val seconds = matcher.groupValues[11].toInt()

        val hoursStr = timeFunc(hours, "h")
        val minutesStr = timeFunc(minutes, "m")
        val secondsStr = timeFunc(seconds, "s")

        return FormatResult(
            " &a&lActive Contest\n &6$crop1, $crop2, $crop3\n&eNext: \n &6$crop4, $crop5, $crop6\n &eIn&f$hoursStr$minutesStr$secondsStr",
            botText = true
        )
    }

    private fun contest4Handler(originalText: String, matcher: MatchResult): FormatResult {
        val crop1 = matcher.groupValues[1]
        val crop2 = matcher.groupValues[2]
        val crop3 = matcher.groupValues[3]
        val hours = matcher.groupValues[4].toInt()
        val minutes = matcher.groupValues[5].toInt()
        val seconds = matcher.groupValues[6].toInt()

        val hoursStr = timeFunc(hours, "h")
        val minutesStr = timeFunc(minutes, "m")
        val secondsStr = timeFunc(seconds, "s")

        return FormatResult(
            " &e&lNext:\n &6$crop1, $crop2, $crop3\n &eIn&f$hoursStr$minutesStr$secondsStr",
            botText = true
        )
    }

    private fun bestiaryHandler(originalText: String, matcher: MatchResult): FormatResult {
        val maxPerPage = CONFIG_I.botCategory.getLineCount()

        val bestiary = matcher.groupValues[1]
        val user = matcher.groupValues[2]
        val profile = matcher.groupValues[3]
        val message = matcher.groupValues[4]

        // A single, more robust pattern to capture all data points at once
        val bestiaryDataPattern = Regex("""(\w[\w\s]*?) (\d+)/(\d+)(?: \(([\d.]+)\))?""")

        val prefix = parse("\n &6&l$bestiary bestiary - &f&l$user (&f&l$profile)&6&l\n ")
        val pages = mutableListOf<Text>()
        val currentPageContent = StringBuilder()
        var entriesOnPage = 0

        // Use the single pattern on the whole message
        for (match in bestiaryDataPattern.findAll(message)) {
            if (entriesOnPage == maxPerPage) {
                pages.add(parse(currentPageContent.toString()))
                currentPageContent.setLength(0) // More efficient than new StringBuilder()
                entriesOnPage = 0
            }

            val name = match.groupValues[1].trim()
            val current = match.groupValues[2]
            val max = match.groupValues[3]
            val kdrString = match.groupValues[4] // Can be empty

            if (entriesOnPage != 0) {
                currentPageContent.append("\n")
            }

            currentPageContent.append(" &e&l$name &f&l$current&e&l/&f&l$max")

            if (kdrString.isNotEmpty()) {
                val kdr = kdrString.toDouble()
                val color = when (kdr) {
                    in 1.00..Double.MAX_VALUE -> ColorCode.GREEN.getMcCode()
                    in 0.66..1.00 -> ColorCode.YELLOW.getMcCode()
                    in 0.33..0.66 -> ColorCode.GOLD.getMcCode()
                    else -> ColorCode.RED.getMcCode()
                }
                currentPageContent.append(" &e&l($color&l$kdrString&e&l)")
            } else {
                currentPageContent.append(" &e&l(&a&lPro&e&l)")
            }
            entriesOnPage++
        }

        // Add the final page if it has any content
        if (currentPageContent.isNotEmpty()) {
            if (!pages.isEmpty()) {
                while (entriesOnPage < maxPerPage) {
                    currentPageContent.append("\n")
                    entriesOnPage += 1
                }
            }
            pages.add(parse(currentPageContent.toString()))
        }

        val titles = mutableListOf<Text>()
        for (i in 1..pages.size) {
            titles.add(parse("&6Page ($i/${pages.size})"))
        }

        return FormatResult(
            pages, titles, TextColor.fromFormatting(Formatting.DARK_AQUA)!!,
            TextColor.fromFormatting(Formatting.GRAY)!!, prefix, botText = true
        )
    }

    private fun bestiary2Handler(originalText: String, matcher: MatchResult): FormatResult {
        val mob = matcher.groupValues[1]
        val user = matcher.groupValues[2]
        val profile = matcher.groupValues[3]
        val num = matcher.groupValues[4]

        val str = if (num.toInt() > 0) "&a&lPro" else "&4&l0"

        return FormatResult(
            "&f&l$user (&f&l$profile)&6&l:\n &6&l$mob - &f&l$num&e&l/&f&l0 &e&l($str&e&l)",
            botText = true
        )
    }

    private fun collectionHandler(originalText: String, matcher: MatchResult): FormatResult {
        val maxPerPage = CONFIG_I.botCategory.getLineCount()

        val skill = matcher.groupValues[1]
        val user = matcher.groupValues[2]
        val profile = matcher.groupValues[3]
        val message = matcher.groupValues[4]

        // 1. Replaced the two patterns with a single, more direct one.
        // This pattern captures all necessary groups in one pass.
        val dataPattern = Regex("""([\w\s]*) (\d+)/(\d+) \(([^)]+)\)""")

        val prefix = parse("\n &6&l${capitalizeFirstLetter(skill)} collections - &f&l$user (&f&l$profile)&6&l\n ")

        val pages = mutableListOf<Text>()
        val currentPageContent = StringBuilder()
        var entriesOnPage = 0

        for (match in dataPattern.findAll(message)) {
            if (entriesOnPage == maxPerPage) {
                pages.add(parse(currentPageContent.toString()))
                currentPageContent.setLength(0)
                entriesOnPage = 0
            }

            val name = match.groupValues[1].trim()
            val current = match.groupValues[2].toInt()
            val max = match.groupValues[3].toInt()
            val progress = match.groupValues[4]

            if (entriesOnPage != 0) {
                currentPageContent.append("\n")
            }

            val median = Math.floorDiv(max, 4)
            val numberColor = when {
                current == max -> ColorCode.GREEN
                current > median -> ColorCode.GOLD
                else -> ColorCode.RED
            }

            val formattedProgress = progress.replace("/", "&e&l/&f&l")

            currentPageContent.append(
                " &e&l$name ${numberColor.getMcCode()}&l$current&e&l/${numberColor.getMcCode()}&l$max &e&l(&f&l$formattedProgress&e&l)"
            )

            entriesOnPage++
        }

        // This logic for handling the final page remains the same.
        if (currentPageContent.isNotEmpty()) {
            if (!pages.isEmpty()) {
                while (entriesOnPage < maxPerPage) {
                    currentPageContent.append("\n")
                    entriesOnPage++
                }
            }
            pages.add(parse(currentPageContent.toString()))
        }

        // This logic for generating titles remains the same.
        val titles = mutableListOf<Text>()
        for (i in 1..pages.size) {
            titles.add(parse("&6Page ($i/${pages.size})"))
        }

        // The return statement is the same.
        return FormatResult(
            pages, titles, TextColor.fromFormatting(Formatting.DARK_AQUA)!!,
            TextColor.fromFormatting(Formatting.GRAY)!!, prefix, botText = true
        )
    }

    private fun discordHandler(originalText: String, matcher: MatchResult): FormatResult {
        val user = matcher.groupValues[1]
        var message = matcher.groupValues[2]
        val userName: String

        // This logic handles cases where the message contains ": " and the regex captures it as part of the user.
        // It correctly reassembles the user and message parts.
        if (user.contains(": ")) {
            val parts = user.split(": ")
            userName = parts[0]
            val messageBuilder = StringBuilder()
            for (i in 1..<parts.size) {
                messageBuilder.append(parts[i]).append(": ")
            }
            messageBuilder.append(message)
            message = messageBuilder.toString()
        } else {
            userName = user
        }

        val discord = CONFIG_I.discordCategory
        val finalMessage = parse(
            "${discord.nameColor.get().getMcCode()}$userName${discord.messageColor.get().getMcCode()}: "
        )

        val parts = message.split(" ")
        var starting = true
        for (part in parts) {
            if (part.isEmpty()) {
                continue
            }

            val linkMatcher = LINK_PATTERN.matchEntire(part)
            //Treat links
            val formattedPart = if (linkMatcher != null) {
                formatLink(part, linkMatcher)
            } else {
                parse("${CONFIG_I.discordCategory.messageColor.get().getMcCode()}$part")
            }

            //Add to the end of the message;
            if (!starting) {
                finalMessage.append(" ")
            } else {
                starting = false
            }
            finalMessage.append(formattedPart)
        }

        return FormatResult(finalMessage, discordText = true)
    }

    fun interface SpecialFunction {
        fun run(originalText: String, matcher: MatchResult): FormatResult
    }
}